src/rust-crypto/rust-crypto.ts

```diff
@@ -755,7 +755,8 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
 
         const secretStorageKeyObject = await this.secretStorage.addKey(
             SECRET_STORAGE_ALGORITHM_V1_AES,
-            secretStorageKey.keyInfo,
+            // add the private key in the options in order to encrypt the zero secret for key checks.
+            { ...secretStorageKey.keyInfo, key: secretStorageKey.privateKey },
```

Comment:
This is basically the fix, but I wonder if the private key should not already be in that `keyInfo`? So maybe should be done by `createSecretStorageKey`. @richvdh I think you are more familiar with that code

Reply:
The place where legacy is doing similar https://github.com/matrix-org/matrix-js-sdk/blob/13c7e0ebda88d3676c8bc403fa4c5bd88a17023a/src/crypto/index.ts#L983

Reply:
> So maybe should be done by `createSecretStorageKey`

Ugh. Yes. In an ideal world, it should. However, in practice, I think it's more complicated.

Currently we have:

```ts
export interface GeneratedSecretStorageKey {
    keyInfo?: AddSecretStorageKeyOpts;
    /** The raw generated private key. */
    privateKey: Uint8Array;
    /** The generated key, encoded for display to the user per https://spec.matrix.org/v1.7/client-server-api/#key-representation. */
    encodedPrivateKey?: string;
}
```

and

```ts
export interface AddSecretStorageKeyOpts {
    pubkey?: string;
    passphrase?: PassphraseInfo;
    name?: string;
    key?: Uint8Array;
}
```

... which means that the private key is duplicated.

I think this structure is needed because there are some flows (notably when the 4S key comes from some customisation of matrix-react-sdk) when the key does not itself need storing in 4S) - so `keyInfo` is undefined. (Of course, that doesn't work with rust crypto at the moment, because of the check on line 752, but that's the idea.)

My suggestion would be:

 * *Relax* the definition of `GeneratedSecretStorageKey` so that it doesn't include the full `AddSecretStorageKeyOpts` structure:
   ```ts
   export interface GeneratedSecretStorageKey {
       keyInfo?: {
           passphrase?: PassphraseInfo;
           name?: string;
       },
       /** The raw generated private key. */
       privateKey: Uint8Array;
       /** The generated key, encoded for display to the user per https://spec.matrix.org/v1.7/client-server-api/#key-representation. */
       encodedPrivateKey?: string;
   }
   ```

   * Make `key` a **required** property of `AddSecretStorageKeyOpts`, since there should be no reason to call `secretStorage.addKey` without setting it. I also think that `pubkey` is unused and should be removed, or at least deprecated.
   
   *  Whereever we try to build an `AddSecretStorageKeyOpts` from a `GeneratedSecretStorageKey`, copy the field across.
   
   * Make sure we *document* the behaviour of these properties. Half the reason we have this mess is that `AddSecretStorageKeyOpts.key` lacks any documentation about wtf it is.